### PR TITLE
tests: Replace counting_illegal_callbacks with CHECK_ILLEGAL_VOID

### LINKED ISSUE #1167: Setting counting_illegal_callback may hide failing tests
The following code snippet very roughly resembles what happens in the libsecp test harness and illustrates the issue:

```C
/* in the tests we have access to a global context */
static secp256k1_context *ctx = NULL;

void test_foo() {
  int ecount;
  context_set_error_callback(ctx, counting_illegal_callback_fn, &ecount);
  /* do some tests with ecount */
}
void test_bar() {
  /* we don't set the counting_illegal_callback here because we don't want to test that here */
  some_function(ctx);
}
void main() {
  test_foo();
  test_bar();
}
```

The code is fine, until one day `some_function(ctx)` results in the illegal callback being called.
Then we'd want the test to fail but instead what happens is that some stack region formerly known as `ecount` is modified, which does not necessarily result in a crash.

One solution would be to never add the `counting_illegal_callback` to the global context and instead create a local context for counting.

### PR DESCRIPTION
**Pull request by jonasnick**
Fixes #1167

### CODE REVIEW

File: src/tests.c [hunk 1]
```diff
@@ -6712,34 +6590,30 @@ static void run_pubkey_comparison(void) {
...
-        CHECK(ecount == 5);
-        CHECK(secp256k1_ec_pubkey_cmp(CTX, &pk2, &pk_tmp) > 0);
-        CHECK(ecount == 6);
+        CHECK_ILLEGAL_VOID(CTX, CHECK(secp256k1_ec_pubkey_cmp(CTX, &pk_tmp, &pk2) < 0));
+        {
+            int32_t ecount = 0;
+            secp256k1_context_set_illegal_callback(CTX, counting_illegal_callback_fn, &ecount);
+            CHECK(secp256k1_ec_pubkey_cmp(CTX, &pk_tmp, &pk_tmp) == 0);
+            CHECK(ecount == 2);
+            secp256k1_context_set_illegal_callback(CTX, NULL, NULL);
```
Comment (real-or-random) [inline src/tests.c:6612]: Is it a bug that `secp256k1_ec_pubkey_cmp` calls the callback twice?

I mean, we never do this anywhere else, even if multiple arguments are illegal. In particular, `ARG_CHECK` makes the function return immediately on failure.

File: src/modules/extrakeys/tests_impl.h [hunk 2]
```diff
@@ -139,29 +118,26 @@ static void test_xonly_pubkey_comparison(void) {
...
-    CHECK(ecount == 5);
-    CHECK(secp256k1_xonly_pubkey_cmp(CTX, &pk2, &pk1) > 0);
-    CHECK(ecount == 6);
+    CHECK_ILLEGAL_VOID(CTX, CHECK(secp256k1_xonly_pubkey_cmp(CTX, &pk1, &pk2) < 0));
+    {
+        int32_t ecount = 0;
+        secp256k1_context_set_illegal_callback(CTX, counting_illegal_callback_fn, &ecount);
+        CHECK(secp256k1_xonly_pubkey_cmp(CTX, &pk1, &pk1) == 0);
+        CHECK(ecount == 2);
+        secp256k1_context_set_illegal_callback(CTX, NULL, NULL);
```
Comment (real-or-random) [inline src/modules/extrakeys/tests_impl.h:138]: (if you change it, also change it here)

File: src/modules/ecdh/tests_impl.h [hunk 3]
```diff
@@ -26,31 +26,19 @@ static int ecdh_hash_function_custom(unsigned char *output, const unsigned char
 
 static void test_ecdh_api(void) {
     /* Setup context that just counts errors */
```
Comment (real-or-random) [inline src/modules/ecdh/tests_impl.h:28]: nit: Comment needs to be removed (or changed)

File: src/tests.c [hunk 4]
```diff
@@ -5962,142 +5916,99 @@ static void run_ec_pubkey_parse_test(void) {
         0xB8, 0x00
     };
     unsigned char sout[65];
-    unsigned char shortkey[2];
+    unsigned char shortkey[2] = { 0 };
```
Comment (real-or-random) [inline src/tests.c:5919]: nit:

```suggestion
    unsigned char shortkey[2] = { 0x03, 0x05 };
```

0x0500000000000000000000000000000000000000000000000000000000000000 is a valid x-coord with odd y-coord

**Review by real-or-random — COMMENTED** [review]
utACK mod nits


This makes the test code much cleaner.

I suggest (this PR or followup)
 * changing `secp256k1_ec_pubkey_cmp` to call a callback at most once (see review)
 * adding a macro `CHECK_ERROR` for the `error_callback` (which should probably involve renaming `counting_illegal_callback` to just `counting_callback` -- the callback doesn't care what it's used for)

File: src/tests.c (hunk 4 above)
Comment (jonasnick) [reply src/tests.c:5919]: I'm not sure how this would improves the tests.

File: src/tests.c (hunk 1 above)
Comment (jonasnick) [reply src/tests.c:6612]: It's intentional, see the comment in `pubkey_cmp`:

```
        /* If the public key is NULL or invalid, ec_pubkey_serialize will call
         * the illegal_callback and return 0. In that case we will serialize the
         * key as all zeros which is less than any valid public key. This
         * results in consistent comparisons even if NULL or invalid pubkeys are
         * involved and prevents edge cases such as sorting algorithms that use
         * this function and do not terminate as a result. */
```

In particular, returning 0 immediately in pubkey_cmp if parsing fails does not signal an error to the color, but that both public keys are identical.

File: src/modules/ecdh/tests_impl.h (hunk 3 above)
Comment (jonasnick) [reply src/modules/ecdh/tests_impl.h]: fixed

**Comment by jonasnick** [comment]
> - adding a macro CHECK_ERROR for the error_callback (which should probably involve renaming counting_illegal_callback to just counting_callback -- the callback doesn't care what it's used for)

done

File: src/tests.c (hunk 4 above)
Comment (real-or-random) [reply src/tests.c:5911]: Yeah sorry, I didn't think it through entirely.

File: src/tests.c (hunk 1 above)
Comment (real-or-random) [reply src/tests.c]: Oh, crazy. I still think we could get the best of both worlds and retain the current all-zeros behavior and at the same time call the callback only once. But that's better done in a separate PR then.

**Review by real-or-random — APPROVED** [review]
utACK fb5eb75a89598394f615eb55f0757e7c35e420ba

**Review by siv2r — COMMENTED** [review]
The `error_callback` and `illegal_callback` are being set to `NULL` at the end of `test_keypair` fn (in `extrakeys/tests_impl.h`). Shouldn't we remove these too?

Github doesn't allow me to comment on these lines. So, here's the link: https://github.com/bitcoin-core/secp256k1/pull/1390/files#diff-07d311190e50e344466bc33ed56c60d2a15f4e8221bfc97792e1cd441b8f3e19L443-L444

**Review by siv2r — COMMENTED** [review]
ACK fb5eb75, the tests locally pass on my machine (for all four commits).

**Comment by real-or-random** [comment]
needs rebase :)

**Comment by jonasnick** [comment]
rebased

**Comment by jonasnick** [comment]
@siv2r 

> The error_callback and illegal_callback are being set to NULL at the end of test_keypair fn (in extrakeys/tests_impl.h).

Good catch, fixed.

**Comment by siv2r** [comment]
reACK 7030364 (tests pass locally)

**Pull request merged**